solr/solrj/src/java/org/apache/solr/common/cloud/DocCollection.java

```diff
@@ -488,4 +494,27 @@ public interface CollectionStateProps {
     String SHARDS = "shards";
     String PER_REPLICA_STATE = "perReplicaState";
   }
+
+  public static class PrsSupplier implements Supplier<PerReplicaStates> {
+
+    private volatile PerReplicaStates prs;
+
+    private Supplier<PerReplicaStates> supplier;
+
+    public PrsSupplier(Supplier<PerReplicaStates> supplier) {
+      this.supplier = supplier;
+    }
+
+    public PrsSupplier(PerReplicaStates prs) {
+      this.prs = prs;
+    }
+
+    @Override
+    public PerReplicaStates get() {
+      if (prs == null) {
```

Comment:
what if multiple threads call this same time? do we see this can be fetched twice/more ?

Reply:
yes. but it is OK as it doesn't cost much

Reply:
Not sure it justify lazy loading. Better to initialize this in DocCollection'ctor, keep it simple. 
It seems like you want some delay in fetching prs state, if so we should consider that problem separately.

Reply:
this object is constructed before `DocCollection` is constructed

Reply:
Sure. I think we want to avoid `LazyPrsSupplier` interface. So that we can construct immutable `DocCollection` object. I'm happy to file separate jira for it.